File: 2197-decode-the-slanted-ciphertext/decode-the-slanted-ciphertext.java
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        final int N = encodedText.length();

        int cols = N / rows;
        StringBuilder sb = new StringBuilder();
        for (int k = 0; k < cols; k++) {
            for (int i = 0, j = k; i < rows && j < cols; i++, j++) {
                sb.append(encodedText.charAt(i * cols + j));
            }
        }
        while (!sb.isEmpty() && sb.charAt(sb.length() - 1) == ' ') {
            sb.deleteCharAt(sb.length() - 1);
        }

        return sb.toString();
    }
}

File: 2197-decode-the-slanted-ciphertext/decode-the-slanted-ciphertext.cpp
class Solution {
public:
    string decodeCiphertext(string encodedText, int rows) {
        const int N = encodedText.size();

        string result;
        int cols = N / rows;
        for (int k = 0; k < cols; k++) {
            for (int i = 0, j = k; i < rows && j < cols; i++, j++) {
                result.push_back(encodedText[i * cols + j]);
            }
        }
        while (!result.empty() && isspace(result.back())) {
            result.pop_back();
        }

        return result;
    }
};